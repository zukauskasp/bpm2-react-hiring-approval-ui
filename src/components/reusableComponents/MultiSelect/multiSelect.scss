.formControl {
    margin-bottom: 24px;
}

.inputWrapper {
    width: 280px;
    font-family: 'Danske Text v2';
    font-weight: 700;
    border: 1px solid #ced4da;
    background-color: #fff;
    border-radius: .25rem;
    padding: 1px;
    display: flex;
    flex-wrap: wrap;
    z-index: 1;

    &:hover {
        border-color: #40a9ff;
    }

    &.focused {
        border-color: #40a9ff;
    }

    & input {
        font-size: 14px;
        height: 48px;
        line-height: 48px;
        padding: 2px 6px;
        width: 0;
        flex-grow: 1;
        border: 0;
        outline: 0;
    }
}

label {
    font-family: 'Danske Text v2';
    display: inline-block;
    margin-bottom: 8px;
    font-size: 14px;
    color: #003755;
}

.tag {
    display: flex;
    align-items: center;
    height: 36px;
    line-height: 30px;
    margin: 6px 3px 0 3px;
    background-color: #C5C8CB;
    border: 1px solid #e8e8e8;
    border-radius: 2px;
    box-sizing: content-box;
    padding: 0 4px 0 10px;
    outline: 0;
    overflow: hidden;
    color: #003755;
    font-size: 12px;

    &:focus {
        border-color: #40a9ff;
        background-color: #e6f7ff;
    }

    & span {
        overflow: hidden;
        white-space: nowrap;
        text-overflow: ellipsis;
    }
}

.rootElement {
    position: relative;
    width: 280px;

    .listBoxWrapper {
        z-index: 3;

        &:after {
            content: "\A";
            position: absolute;
            transform: translate(50%, 160%);
            bottom: 0%;
            left: 80%;
            width: 0;
            height: 0;
            border-left: 10px solid transparent;
            border-right: 10px solid transparent;
            border-bottom: 10px solid #e8e8e8;
        }

        &:before {
            content: "\A";
            position: absolute;
            transform: translate(50%, 180%);
            bottom: 0%;
            left: 81%;
            width: 0;
            height: 0;
            border-left: 9px solid transparent;
            border-right: 9px solid transparent;
            border-bottom: 9px solid #fff;
            z-index: 4;

        }

        .listBox {
            width: 280px;
            padding: 0;
            top: 100%;
            list-style: none;
            background-color: #fff;
            overflow: auto;
            max-height: 250px;
            border-radius: 4px;
            z-index: 2;
            position: absolute;
            border: 1px solid #e8e8e8;

            & li {
                font-size: 14px;
                height: 48px;
                padding: 5px 12px;
                margin: 1px;
                display: flex;
                align-items: center;
                color: #003755;

                & span {
                    flex-grow: 1;
                }
            }

            & li[aria-selected='true'] {
                background-color: #C5C8CB;
            }

            & li[data-focus='true'] {
                background-color: #C5C8CB;
                cursor: pointer;
            }
        }
    }
}