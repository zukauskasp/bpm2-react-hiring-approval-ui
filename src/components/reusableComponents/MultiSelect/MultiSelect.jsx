import React from "react";
import './multiSelect.scss';
import PropTypes from 'prop-types';
import Tooltip from "../Tooltip/Tooltip";
import useAutocomplete from '@material-ui/lab/useAutocomplete';
import styled from 'styled-components';
import Icon from "../Icon/Icon";


const MultiSelect = ({ errorMessage, label, list, onChange }) => {
    const {
        getRootProps,
        getInputProps,
        getTagProps,
        getListboxProps,
        getOptionProps,
        groupedOptions,
        value,
        focused,
        setAnchorEl,
    } = useAutocomplete({
        defaultValue: [],
        multiple: true,
        options: list,
        onChange: onChange,
        getOptionLabel: option => option.value,
    });
    return (
        <div className="rootElement">
            <div {...getRootProps()} >
                {label ? <label >{label}</label> : null}
                <div ref={setAnchorEl} className={`inputWrapper ${focused ? 'focused' : ''}`} >
                    {value.map((option, index) => (
                        <Tag label={option.name} {...getTagProps({ index })} />
                    ))}
                    <input {...getInputProps()} />
                </div>
            </div>
            {errorMessage ? <Tooltip text={errorMessage} messageType='error' /> : null}
            {groupedOptions.length > 0 ? (
                <div className="listBoxWrapper" >
                    <ul className="listBox" {...getListboxProps()}   >
                        {groupedOptions.map((option, index) => (
                            < li {...getOptionProps({ option, index })} >
                                <span>{option.name}</span>
                                <Icon
                                    icon="remove"
                                    color={value.indexOf(option) > -1 ? "white" : 'dark'}
                                    className="custom-list-icon"
                                />
                            </li>
                        ))}
                    </ul>
                </div>
            ) : null
            }
        </div >
    );
}

MultiSelect.propTypes = {
    errorMessage: PropTypes.string,
    label: PropTypes.string,
    onChange: PropTypes.func.isRequired,
    list: PropTypes.arrayOf(PropTypes.exact({
        name: PropTypes.string,
        value: PropTypes.string
    })).isRequired,
}

MultiSelect.defaultProps = {
    errorMessage: "",
    label: ""
}


export default MultiSelect;


const Tag = styled(({ label, onDelete, ...props }) => (
    <div {...props}>
        <span>{label}</span>
        <Icon icon="remove" alt="" onClick={() => onDelete()} style={{ width: '1.2em', margin: '5px' }} />
    </div>
))`
  display: flex;
  align-items: center;
  height: 36px;
  line-height: 30px;
  margin: 6px 3px 0 3px;
  background-color: #C5C8CB;
  border: 1px solid #e8e8e8;
  border-radius: 2px;
  box-sizing: content-box;
  padding: 0 4px 0 10px;
  outline: 0;
  overflow: hidden;
  color: #003755;
  font-size: 12px;

  &:focus {
    border-color: #40a9ff;
    background-color: #e6f7ff;
  }

  & span {
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
  }

  & svg {
    font-size: 12px;
    cursor: pointer;
    padding: 4px;
  }
`;
