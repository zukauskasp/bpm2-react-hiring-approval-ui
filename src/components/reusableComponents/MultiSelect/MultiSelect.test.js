import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';

import MultiSelect from './MultiSelect';

describe('MultiSelect', () => {

    let wrapper;
    let mount;
    let shallow;

    beforeAll(() => {
        shallow = createShallow();
        mount = createMount();
    });

    const list = [
        { name: "one", value: "1" },
        { name: "two", value: "2" },
        { name: "three", value: "3" },
    ]
    const props = {
        list: list,
        label: "Select number",
        onChange: () => { },
        errorMessage: "Error message"
    }

    beforeEach(() => {
        wrapper = shallow(<MultiSelect {...props} />);
    })

    afterAll(() => {
        mount.cleanUp();
      });

    it('expect to match latest snapshot', () => {
        expect(shallow(<MultiSelect {...props} />).debug()).toMatchSnapshot();
    });

    it('expect to render MultiSelect', (() => {
        expect(wrapper.length).toEqual(1);
    }));

    it('expect to render correct props', () => {
        expect(mount(<MultiSelect {...props} />).props()).toEqual(props);
    });

    it('expect to render default props', () => {
        const mounted = mount(<MultiSelect list={props.list} onChange={props.onChange} />);

        expect(mounted.props().errorMessage).toEqual("");
        expect(mounted.props().label).toEqual("");
    });

    it('expect to render list', () => {
        const mounted = mount(<MultiSelect {...props} />);

        expect(mounted.prop('list').length).toEqual(list.length);
    });

    it('expect to render label and error message correctly', () => {
        let mounted = mount(<MultiSelect list={props.list} onChange={props.onChange} />);
        expect(mounted.prop('label')).toEqual('');
        expect(mounted.prop('errorMessage')).toEqual('');

        mounted = mount(<MultiSelect {...props} />);
        expect(mounted.prop('label')).toEqual(props.label);
        expect(mounted.prop('errorMessage')).toEqual(props.errorMessage);
    });

}) 