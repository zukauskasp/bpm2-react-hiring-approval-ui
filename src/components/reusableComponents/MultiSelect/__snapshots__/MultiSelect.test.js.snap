// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect expect to match latest snapshot 1`] = `
"<div className=\\"rootElement\\">
  <div aria-owns={{...}} role=\\"combobox\\" aria-expanded={false} onKeyDown={[Function]} onMouseDown={[Function: handleMouseDown]} onClick={[Function: handleClick]}>
    <label>
      Select number
    </label>
    <div className=\\"inputWrapper \\">
      <input id={[undefined]} value=\\"\\" onBlur={[Function: handleBlur]} onFocus={[Function: handleFocus]} onChange={[Function: handleInputChange]} onMouseDown={[Function: handleInputMouseDown]} aria-activedescendant={{...}} aria-autocomplete=\\"list\\" aria-controls={{...}} autoComplete=\\"off\\" autoCapitalize=\\"none\\" spellCheck=\\"false\\" />
    </div>
  </div>
  <Message text=\\"Error message\\" messageType=\\"error\\" position=\\"top-left\\" />
</div>"
`;
